src/runtime_strings.js

```diff
@@ -63,12 +63,18 @@ var UTF8Decoder = typeof TextDecoder !== 'undefined' ? new TextDecoder('utf8') :
 #endif // TEXTDECODER
 #endif // TEXTDECODER == 2
 
-function UTF8ArrayToString(u8Array, idx) {
+/**
+ * @param {number} idx
+ * @param {number=} endIdx
+ * @return {string}
+ */
+function UTF8ArrayToString(u8Array, idx, endIdx) {
 #if TEXTDECODER
   var endPtr = idx;
   // TextDecoder needs to know the byte length in advance, it doesn't stop on null terminator by itself.
   // Also, use the length info to avoid running tiny strings through TextDecoder, since .subarray() allocates garbage.
-  while (u8Array[endPtr]) ++endPtr;
+  // (As a tiny code save trick, compare endPtr against endIdx using a negation, so that undefined means Infinity)
```

Comment:
trick is nice for code size, but for speed it may be harmful, as the VM will see different types here, and not just the simple case of i32s. is the code size worth that?

Reply:
The reason I am squeezing out every single byte of these "hello console.log" and "hello webgl" applications is that these contain functions that are present practically in every program compiled wiht Emscripten. Having those be hyperoptimized does seem very worthwhile to me, and it helps build that impression that "Emscripten is not bloat" when they can do the minimal test apps and read through every line to understand how the generated code is structured.

That aside, there's two aspects to the performance question. First, this function is called with the endIdx argument only from GL.getSource() function. If user is not doing GL (glShaderSource), then at compilation time Closure sees that the endIdx is never passed, and it optimizes the function so that it does not take endIdx as input. This does already happen, although due to the above linked bug, the transformation it is not yet optimal in size.

glShaderSource API absolutely does require this endIdx-provided form of the string marshalling, unfortunately no way to implement it without. I'd hate to create a second UTF8 string marshalling function just for this one place GL.getSource() to call, so this was the shortest way to support both uses that I could think.

Second, my understanding is that such JIT bailouts due to unexpected types will occur only when the types change at runtime dynamically. So (currently) in WebGL utilizing programs, the type variance will only trigger at times when glShaderSource() is called, which should be in any program's cold path. Only if application developer heavily adds in uses of this new form of UTF8ToString with endIdx provided, this kind of random mix of endIdx==undefined vs endIdx==i32 type variance would occur. So I am not worried (today). And it is not certain if such type variance will end up with degraded performance, and by how much. So I think I'd like to take the code size win here now, and if someone then runs into a case that shows the performance loss by unoptimizable JIT, then accept the code size impact by filling in an endIdx of Infinity or similar. (although that will require creating two string marshalling functions I think, at which point these functions should definitely live in a JS library)

Reply:
Added a note. Also realized that the function `stringToUTF8` for the opposite direction did already take in a max length (`maxBytesToWrite`), and not a `endPtr`, so converted this to also take in `maxBytesToRead` instead of `endPtr` so that the function signatures are symmetric in both directions. I'll assume that's an obviously good change, so I'll land this so that I can layer other PRs on top of this already today.